### ISSUE TITLE
Decode fails on key name with an underscore

### ISSUE THREAD

**Issue by samadhicsec** [mitchellh/mapstructure#88]
See attached file for test case with the problem that I am seeing.
[test.go.txt](https://github.com/mitchellh/mapstructure/files/1138329/test.go.txt)

Basically I want to convert part of a JSON string to a struct.  In the example this is the 'key' section.  To do the conversion I unmarshal the JSON to an unknown map, extract part of the map I am interested in, and then use Decode to convert that map to a struct.

The example has 2 (hopefully) identical functions processing slightly different JSON.  In the first function a nested JSON key has an underscore in it, and in the second function the same nested key does not have an underscore in it.

The output of the first function shows that Decode loses the value of that key with the underscore.  The output of the second function shows that Decode does not lose the value of that key.  I have tried to create the example so that the only meaningful difference between these 2 functions is the underscore in the key name.

Apologies for not finding the actual bug (assuming there is one), hopefully you can easily find it and fix it.

**Comment by weilixu7** [comment]
I met the same problem, it's failed when structs with underline tag

**Comment by kehindesalaam** [comment]
@samadhicsec did you later find a solution to this problem

**Comment by craig-duffin** [comment]
Would be great if there was a solution to this issue

**Comment by kehindesalaam** [comment]
I already fixed it  in #99

**Comment by samadhicsec** [comment]
So I don't think there is a bug here to fix, instead I failed to set the TagName in the DecoderConfig to "json" (and the default value is "mapstructure").  Once I did that, so

`config := &mapstructure.DecoderConfig{
	    TagName: "json",
	    Result: &vsv,
}`
`decoder, err := mapstructure.NewDecoder(config)`

Then the code will look for the struct tag name instead of the struct element name, and the correct key name will be found in the map and the correct value copied to the struct.

An end to end json example might be a useful addition to your documentation, as it is likely the most common use case.  In my particular case I was auto-generating the struct definitions from json using https://github.com/gnewton/chidley, so all the tag names were pre-set to "json".
